include/albatross/src/indexing/apply.hpp

```diff
@@ -54,6 +40,70 @@ auto apply(const std::vector<ValueType> &xs, const ApplyFunction &f) {
   return output;
 }
 
+// Map
+
+template <
+    template <typename...> class Map, typename KeyType, typename ValueType,
+    typename ApplyFunction,
+    typename ApplyType = typename details::key_value_apply_result<
+        ApplyFunction, KeyType, ValueType>::type,
+    typename std::enable_if<details::is_valid_key_value_apply_function<
+                                ApplyFunction, KeyType, ValueType>::value &&
+                                std::is_same<void, ApplyType>::value,
+                            int>::type = 0>
+void apply(const Map<KeyType, ValueType> &map, const ApplyFunction &f) {
+  for (const auto &pair : map) {
+    f(pair.first, pair.second);
+  }
+}
+
+template <
+    template <typename...> class Map, typename KeyType, typename ValueType,
+    typename ApplyFunction,
+    typename ApplyType = typename details::key_value_apply_result<
+        ApplyFunction, KeyType, ValueType>::type,
+    typename std::enable_if<details::is_valid_key_value_apply_function<
+                                ApplyFunction, KeyType, ValueType>::value &&
+                                !std::is_same<void, ApplyType>::value,
+                            int>::type = 0>
+auto apply(const Map<KeyType, ValueType> &map, const ApplyFunction &f) {
+  Grouped<KeyType, ApplyType> output;
+  for (const auto &pair : map) {
+    output.emplace(pair.first, f(pair.first, pair.second));
+  }
+  return output;
+}
+
+template <template <typename...> class Map, typename KeyType,
+          typename ValueType, typename ApplyFunction,
+          typename ApplyType = typename details::value_only_apply_result<
+              ApplyFunction, ValueType>::type,
+          typename std::enable_if<details::is_valid_value_only_apply_function<
+                                      ApplyFunction, ValueType>::value &&
+                                      !std::is_same<void, ApplyType>::value,
+                                  int>::type = 0>
+auto apply(const Map<KeyType, ValueType> &map, const ApplyFunction &f) {
+  Grouped<KeyType, ApplyType> output;
+  for (const auto &pair : map) {
+    output.emplace(pair.first, f(pair.second));
+  }
+  return output;
+}
+
+template <template <typename...> class Map, typename KeyType,
+          typename ValueType, typename ApplyFunction,
+          typename ApplyType = typename details::value_only_apply_result<
+              ApplyFunction, ValueType>::type,
+          typename std::enable_if<details::is_valid_value_only_apply_function<
+                                      ApplyFunction, ValueType>::value &&
+                                      std::is_same<void, ApplyType>::value,
+                                  int>::type = 0>
+auto apply(const Map<KeyType, ValueType> &map, const ApplyFunction &f) {
```

Comment:
Return type should probably be explicitly `void`

Reply:
good call